File: PS_Study/BOJ/1932.cpp
#include <cstdio>
#include <algorithm>
using namespace std;

int dp[501];

int main(void)
{
	int n, val, i, j;
	freopen("input.txt", "rt", stdin);
	scanf("%d", &n);

	for (i = 1; i <= n; i++) {
		for (j = i; j >= 1; j--) {
			scanf("%d", &val);
			dp[j] = max(dp[j] + val, dp[j - 1] + val);
		}
	}

	printf("%d\n", *max_element(dp+1, dp+1+n ));

	return 0;
}

File: PS_Study/BOJ/1932.py
N = int(input())
arr = [[0 for _ in range(N+1)] for __ in range(N+1)]
dp = [[0 for _ in range(N+1)] for __ in range(N+1)]

for i in range(1, N+1):
    tmp = list(map(int, input().split()))
    for j in range(1, i+1):
        arr[i][j] = tmp[j-1]

for i in range(1, N+1):
    for j in range(1, N+1):
        dp[i][j] = max(dp[i-1][j-1], dp[i-1][j]) + arr[i][j]

print(max(dp[-1]))